### PR TITLE
lint #1674: replace struct name with `Self` when applicable

### PR DESCRIPTION
**Pull request by montrivo**
Hello, 

This tries to implement issue #1674. Let me know if it needs modifications in order to be merged.

Cheers

### CODE REVIEW

File: clippy_lints/src/use_self.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    item_path: &'a Path,
+    cx: &'a LateContext<'a, 'tcx>,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for UseSelfVisitor<'a, 'tcx> {
+    fn visit_path(&mut self, path: &'tcx Path, _id: NodeId) {
+        if self.item_path.def == path.def &&
+           path.segments
+            .last()
+            .expect("segments should be composed of at least 1 elemnt")
```
Comment (oli-obk) [inline clippy_lints/src/use_self.rs]: elemnt -> element

File: clippy_lints/src/use_self.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            .name
```
Comment (oli-obk) [inline clippy_lints/src/use_self.rs]: You can compare `name` with http://manishearth.github.io/rust-internals-docs/syntax_pos/symbol/keywords/constant.SelfType.html in order to not have to do `as_str` and compare with a string

File: clippy_lints/src/use_self.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            .as_str() != "Self" {
+            span_lint(self.cx,
```
Comment (oli-obk) [inline clippy_lints/src/use_self.rs]: Can you add a suggestion that suggests replacing the explicit name with `Self`?

Grep for `span_suggestion` if you need any examples on how to do it. If you run into any problems don't hesitate to ask

File: tests/ui/use_self.stderr [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+error: repetitive struct name usage. Use `Self` instead.
+  --> $DIR/use_self.rs:16:22
+   |
+16 |         fn test() -> Foo {
+   |                      ^^^
+
+error: repetitive struct name usage. Use `Self` instead.
+  --> $DIR/use_self.rs:17:13
+   |
+17 |             Foo::new()
```
Comment (oli-obk) [inline tests/ui/use_self.stderr]: The span of this one isn't quite right. It should only point to the `Foo` part

Comment (oli-obk) [inline tests/ui/use_self.stderr]: Nevermind, this is a rustc bug. Everything's good

**Comment by oli-obk** [comment]
Thanks. This one will clean up a lot of code.

**Comment by montrivo** [comment]
Thank you for the review!

**Comment by oli-obk** [comment]
@montrivo this introduced some false positives. Do you want to take a crack at them?

Most notably, this triggers on `#[derive(Copy, Clone)]` on a struct (so you need an `is_macro`) check.

But it also triggers on things like `fn foo<'b>(x: &'b str) -> Foo<'b>`, where `Foo<'b>` can't be replaced by `Self`. Maybe just bail out if there are generic arguments for now?

**Comment by montrivo** [comment]
Yes.
Should I add the lint to one of the groups?
What tests exactly did you run to find out about the fasle positives?

2017-08-21 10:26 GMT+02:00 Oliver Schneider <notifications@github.com>:

**Comment by oli-obk** [comment]
I already fixed the `Copy, Clone` issue, since it was bugging me during clippy development. But the other issue is still there.

> What tests exactly did you run to find out about the false positives?

I didn't run any, but I found them inside clippy (the `utils/sugg.rs` file to be exact)

An example is:

```rust
struct Foo<'a>(&'a str);
impl<'a> Foo<'a> {
    // Cannot use `Self` as return type, because the function is actually `fn foo<'b>(s: &'b str) -> Foo<'b>`
    fn foo(s: &str) -> Foo {
        Self(s)
    }
    // cannot replace with `Self`, because that's `Foo<'a>`
    fn foo() -> Foo<'static> {
        Self("foo")
    }
}
```

There are probably more issues once you have generic structs.

I think the only time the suggestion should come in `impl`s, is when it is `impl $something` and the replacement is for `$something` exactly, without any changes in the generic or lifetime parameters. This also applies to `impl SomeTrait for $something`.

I think simply adding a lot of tests with various features like associated types, trait impls setting said associtated types to `Self` and all kinds of generics would go a long way to ensuring we don't break the lint in the future, too.

**Comment by Arnavion** [comment]
@oli-obk The lint also fires for `#[derive(Debug)] struct Foo { }` because the expanded code contains `match *self { Foo { } => ...` Your fix for `Copy` and `Clone` fixes this too, right?

**Comment by montrivo** [comment]
Is `Self` even supported for named tuples?
I get 
```
------------------------------------------
stderr:
------------------------------------------
error[E0423]: expected function, found self type `Self`
  --> tests/ui/use_self.rs:61:13
   |
61 |             Self("foo")
   |             ^^^^ did you mean `Self { /* fields */ }`?
```
with
```rust
mod named_tuples {
    struct Foo(&'static str);
    impl Foo {
        fn new() -> Self {
            Foo("foo")
        }
    }
}
```

**Pull request merged**